### ISSUE TITLE
RC2 - Content in seperate projects

### ISSUE THREAD

**Issue by dazinator** [aspnet/Home#1564]
**TLDR**: How do you serve up static files from referenced projects - and from referenced NuGet packages? In RC1 this was fairly trivial.

## In RC1

If your application referenced a bunch of packages (and projects) - when you build the application, each of the packages / projects that are referenced would remain in the same location, thus when you run the application, you could access info about those projects / packages, including the paths to them on disk, via `ILibraryManager`. This makes it trival to set up `IFileProvider`s for static files middleware to serve content directly from those locations.

Here is an example of some RC1 code that sets up some `IFileProvider`'s to server content from referenced packages (shortened for brevity):

```

             List<IFileProvider> externalFileProviders = new List<IFileProvider>();          
            var referencedLibraries = _libraryManager.GetReferencingLibraries(thisAssemblyName);;
            foreach (var library in referencedLibraries)
            {
                      var libraryPath = library.Path;
                      // Here you would typically do things like see if the package has a 
                      // is a project.json file with a "webroot" property, if it does you'd set up a 
                      // a physical file provider to serve static files from that location within the package. 
                    
                      var contentPath = Path.Combine(libraryPath, "webroot");                      
                      var phsyicalFileProvider = new PhysicalFileProvider(contentPath);
                      externalFileProviders.Add(phsyicalFileProvider);

              }                 

```

## In RC2

In RC2 world, with no `Dnx` - we have [DependencyContext](https://github.com/aspnet/Announcements/issues/149) instead. This accesses metadata that is compiled into assemblies (AFAIK) to give you some package information (dependencies etc) and so it does not utilise `project.json` at runtime. 

Each referenced dependecy (project / package) now has it's ouput copied to your applications bin directory.

Seemingly though there is no convention for having static files (i.e a webroot dir) automatically copied to the output directory. So this means for project references, i'm assuming you now have to get the build process to copy any static files from referenced projects, into your applications output dir and serve from there.. and if its a NuGet package reference rather than a project reference, the story is different - i'm guessing you are stuffed as NuGet packages don't have content copied to output on build, only assemblies.

I have tried using `copyToOutput` in my referenced projects `project.json` file (buildOptions) but this only seems to result in files appearing in that particular projects bin folder, they are not also placed in my main applications output folder which references the project.

Overall this all feels like a big regression from RC1. Even if succeed in getting the project's content files copied to the main applications bin folder on build, does it now mean that we now have to stop and rebuild the application on every change to a static / content file (i.e js file) ? The nice thing about serving the files from their original locations was, for projects in the same solution, you could edit them in place (on the fly) in VS, and refresh the page - no rebuild required..

**Comment by davidfowl** [comment]
There's no "deploy as source" anymore in ASP.NET Core >= RC2. The package layout is no longer preserved in the output format. You'd have to detect what mode you're in and either point to the project directory or the published folder.

**Comment by dazinator** [comment]
> You'd have to detect what mode you're in and either point to the static files middleware to the project directory or the published folder.

If the content is contained within a referenced NuGet package though (We have several NuGet packages that contain MVC components as well as static files like .js files etc) - then how are we supposed to get this content published with our application?

**Comment by dazinator** [comment]
Here is the use case i'm thinking of: Developer A creates some re-usable widget - lets say it's a `ViewComponent` that displays number of open issues on a Github repo. He publishes that to the local company NuGet feed. In other solutions developers wanted to use that widget can pull in that package, and update it when newer versions become available. In their `startup.cs` they can register it with the new `ApplicationParts` system in MVC if necessary. However the widget contains a .js file to make it work. What's the story for how that .js file get's served up in the production app?

**Comment by davidfowl** [comment]
Embed the static content in the assembly as an embedded resource and use an EmbeddedFileProvider to get the static content.

**Comment by dazinator** [comment]
Thanks David. That seems nice to me for referenced NuGet packages.

However for referenced projects, the story remains a little more complicated.

When running the solution from source, you won't want to use embedded file providers for static content in referenced projects, because the workflow of having to build and restart the application when making a .js file change is not productive. However, when running as a published app, the source is not available, so actually using `EmbeddedFileProvider`s for referenced projects static content seems like the easiest solution and nice because it makes your application code consistent with referenced NuGet packages - i.e you are using `EmbeddedFileProvider's` for both referneced packages and referenced projects..

So the easiest route forward I can see, that would still allow a nice workflow, is:

1. Always `embed` static content (i.e typically the wwwroot folder)
2. Use `EmbeddedFileProviders` for all static content by default (referenced projects, and nuget packages)
3. Add some special casing code, so when running your application from source, you create `PhysicalFileProvider`s rather than `EmbeddedFileProvider`s for referenced projects in your solution.

This requires that we are able to tell the difference between an app running as a published app, and an app running from source code. I'm not sure what the best mechanism is for that.. perhaps checking for a solution file? Do you have any suggestions on that?

Thanks for your guidance so far
